inference-engine/src/mkldnn_plugin/nodes/gather_nd.cpp

```diff
@@ -13,75 +15,84 @@ namespace InferenceEngine {
+    static bool isSupportedOperation(const std::shared_ptr<ngraph::Node>& op, std::string& errorMessage) noexcept {
+        try {
+            auto gatherElementsOp = ngraph::as_type_ptr<const ngraph::op::v5::GatherND>(op);
+            if (!gatherElementsOp) {
+                errorMessage = "Node is not an instance of the GatherElements operation from operation set v6.";
+                return false;
+            }
+        } catch (...) {
+            return false;
         }
 
-        Precision indicesPrecision = indices->getTensorDesc().getPrecision();
-        if (indicesPrecision != Precision::I32 &&
-                indicesPrecision != Precision::I16 && indicesPrecision != Precision::U16 &&
-                indicesPrecision != Precision::I8 && indicesPrecision != Precision::U8) {
-            IE_THROW() << _errorPrefix << " has unsupported 'indices' input precision: " << indicesPrecision;
-        }
+        return true;
+    }
 
-        _dataTypeSize = dataPrecision.size();
-        const auto& dataDims = data->getTensorDesc().getDims();
-        const auto& indicesDims = indices->getTensorDesc().getDims();
+    explicit GatherNDImpl(const std::shared_ptr<ngraph::Node>& op) {
+        try {
+            std::string errorMessage;
+            if (!isSupportedOperation(op, errorMessage)) {
+                IE_THROW(NotImplemented) << errorMessage;
+            }
+            _errorPrefix = std::string("Layer GatherND with name '") + op->get_friendly_name() + "'";
 
-        _batchDims = layer->GetParamAsInt("batch_dims", 0);
-        if (_batchDims >= std::min(dataDims.size(), indicesDims.size()))
-            IE_THROW() << _errorPrefix << " has invalid batch_dims attribute: " << _batchDims;
+            if (op->get_input_size() != 2 || op->get_output_size() != 1)
+                IE_THROW() << _errorPrefix << " has invalid number of input/output edges.";
 
-        _batchNum = 1lu;
-        for (size_t i = 0; i < _batchDims; i++) {
-            _batchNum *= indicesDims[i];
-        }
+            Precision inDataPrecision = details::convertPrecision(op->get_input_element_type(_dataIndex));
+            if (inDataPrecision.size() != sizeof(PrecisionTrait<Precision::I32>::value_type) &&
+                    inDataPrecision.size() != sizeof(PrecisionTrait<Precision::I16>::value_type) &&
+                    inDataPrecision.size() != sizeof(PrecisionTrait<Precision::I8>::value_type)) {
+                IE_THROW() << _errorPrefix << " has unsupported 'data' input precision: " << inDataPrecision;
+            }
 
-        _sliceRank = indicesDims[indicesDims.size() - 1];
-        _dataRank = dataDims.size() - _batchDims;
-        if (_sliceRank > _dataRank)
-            IE_THROW() << _errorPrefix << " has invalid inputs shapes.";
+            Precision indicesPrecision = details::convertPrecision(op->get_input_element_type(_indicesIndex));
+            if (indicesPrecision != Precision::I32 &&
```

Comment:
maybe better to use ngraph ```Type``` class method ```is_integral()```?

Reply:
is_float() from IE was added instead

Reply:
what if indicesPrecision will be BOOL?

Reply:
Fixed